src/cow-react/modules/swap/services/safeBundleFlow/index.ts

```diff
@@ -0,0 +1,81 @@
+import { logTradeFlow } from '@cow/modules/trade/utils/logger'
+import { SafeBundleFlowContext } from '@cow/modules/swap/services/types'
+import { PriceImpact } from '@src/hooks/usePriceImpact'
+import { Percent } from '@uniswap/sdk-core'
+import { buildApproveTx } from '@cow/modules/operations/bundle/approve'
+import { buildPresignTx } from '@cow/modules/operations/bundle/presign'
+import { signAndPostOrder } from '@src/utils/trade'
+import { MetaTransactionData } from '@safe-global/safe-core-sdk-types'
+import { getSwapErrorMessage } from '@cow/modules/trade/utils/swapErrorHelper'
+// import { tradeFlowAnalytics } from '@cow/modules/trade/utils/analytics'
+
+const LOG_PREFIX = 'SAFE BUNDLE FLOW'
+
+export async function safeBundleFlow(
+  input: SafeBundleFlowContext,
+  priceImpactParams: PriceImpact,
+  confirmPriceImpactWithoutFee: (priceImpact: Percent) => Promise<boolean>
+): Promise<void> {
+  logTradeFlow(LOG_PREFIX, 'STEP 1: confirm price impact')
+
+  if (priceImpactParams?.priceImpact && !(await confirmPriceImpactWithoutFee(priceImpactParams.priceImpact))) {
+    return
+  }
+
+  try {
+    // For now, bundling ALWAYS includes these 2 steps.
+    // In the feature users will be able to sort/add steps as they see fit
+    logTradeFlow(LOG_PREFIX, 'STEP 2: approval tx')
+    const approveTx = await buildApproveTx({
+      erc20Contract: input.erc20Contract,
+      spender: input.spender,
+      amountToApprove: input.context.trade.inputAmount,
+    })
+    logTradeFlow(LOG_PREFIX, 'STEP 3: post order')
+    const { id: orderId, order: _order } = await signAndPostOrder(input.orderParams).finally(() => {
+      input.callbacks.closeModals()
+    })
+
+    logTradeFlow(LOG_PREFIX, 'STEP 4: ORDER', orderId, _order)
+    logTradeFlow(LOG_PREFIX, 'STEP 4: presign tx')
+    const presignTx = await buildPresignTx({ settlementContract: input.settlementContract, orderId })
+    logTradeFlow(LOG_PREFIX, 'STEP 5: send safe tx')
+    const safeTransactionData: MetaTransactionData[] = [
+      { to: approveTx.to!, data: approveTx.data!, value: '0', operation: 0 },
+      { to: presignTx.to!, data: presignTx.data!, value: '0', operation: 0 },
+    ]
+    // Get safe sdk instance
+    const safeSdk = input.safeSdk
+    const safeTx = await safeSdk.createTransaction({ safeTransactionData, onlyCalls: true })
```

Comment:
Why is `onlyCalls` needed?

Reply:
My understanding from the conversation with the Safe team is that this is what I had to use for doing batch calls via WC.
See the code reference here https://github.com/safe-global/safe-core-sdk/blob/f2e8e82d88d815d7b278f605a125f4cfb2816020/packages/protocol-kit/src/Safe.ts#L377

When I try with `onlyCalls: false` this is the resulting simulation https://dashboard.tenderly.co/public/safe/safe-apps/simulator/42cebc45-fa41-4f0b-ad0b-b5f99727f1ee?trace=0.0.0.4.0.0

Reply:
My first attempt was to do just that: propose a transaction. See this comment where I was discussing with Nick that possibility https://cowservices.slack.com/archives/C043PJXF657/p1683544530087209?thread_ts=1683303573.748569&cid=C043PJXF657

The commented out code is for exactly this, which didn't work of course.

https://github.com/cowprotocol/cowswap/pull/2434/files#diff-ca3f429611c9019673c8bc938d03d5cf47cd916f8ef7df78c0c20d9e3cb777e4R61-R67

Reply:
> The commented out code is for exactly this, which didn't work of course.

This is not what i mean. In the commented code is making a proposal to the API. I think this job will be of the wallet, once the transaction is signed.  

What I'm suggesting is to perform a regular transaction where FROM is the safe, TO is multisend, value is 0, DATA is the multisend params abi-encoding these 2 calls. The internal calls should be delegate calls, so the msg.sender is the safe.

One simple way to do it, is using typechain to generate the method for `multicall` so we can comfortably call it like this https://github.com/cowprotocol/cowswap/blob/c4025003ba8b39a3c9762ca9f442136c92f712a8/src/cow-react/modules/swap/services/swapFlow/steps/presignOrderStep.ts#L25 

I would think that using the connector as you are doing, is also correct if you pass the correct data.

Maybe I'm missing something, I let Nick or Safe crew point it out 🙏

Reply:
> What I'm suggesting is to perform a regular transaction where FROM is the safe, TO is multisend, value is 0, DATA is the multisend params abi-encoding these 2 calls. The internal calls should be delegate calls, so the msg.sender is the safe.

My understanding is that I'm doing that here.

Take this simulation for example: https://dashboard.tenderly.co/public/safe/safe-apps/simulator/560ff74a-606e-4d59-8672-70a7966cc3cf

It comes from the safe `0x5b0abe214ab7875562adee331deff0fe1912fe42` to `0x40a2accbd92bca938b02010e17a5b8929b49130d` (via the Safe fallback handler at `0xb84ef42568a67ab9f27e95e978f85d3ddc0aaf30` I assume)
![image](https://user-images.githubusercontent.com/43217/236839585-a4c843cf-b8c6-4891-a03c-667190dbede3.png)

Except the part of the internal delegate calls, as the multisend contract used doesn't support this.

When using a different multisend though, it's failing as it's not called as a delegate call: https://dashboard.tenderly.co/public/safe/safe-apps/simulator/42cebc45-fa41-4f0b-ad0b-b5f99727f1ee

So I guess I'm back at the first question. How can I call any of these contracts via delegate?

Reply:
But this is not a safe https://goerli.etherscan.io/address/0x5b0abe214ab7875562adee331deff0fe1912fe42

I guess this is a signer inside the safe, and in the simulation there's a signature for this signer. 


I guess the safe is `0x3e5c63644e683549055b9be8653de26e0b4cd36e` and i can see you must be really close, cause here the sender on the multicall is the safe 

<img width="1514" alt="image" src="https://user-images.githubusercontent.com/2352112/236859678-743a11ed-6bb3-412d-8074-86dcc696229e.png">


Inside it, is doing the call to `0x40a2accbd92bca938b02010e17a5b8929b49130d`, which is multisend  (nice!)

I can see the right msg.sender makes it until here the pre-sign call:
<img width="1444" alt="image" src="https://user-images.githubusercontent.com/2352112/236860999-61280a0b-1de0-4660-a99c-162504a0df2e.png">

If i understand it correctly, why it fails, is not because the msg.sender is incorrect, but because the owner is `0xb84ef42568a67ab9f27e95e978f85d3ddc0aaf30`.

Are you passing the right address to pre-sign? the `from` should be `0x3e5c63644e683549055b9be8653de26e0b4cd36e` instead